internal/util/get/get.go

```diff
@@ -16,55 +16,71 @@
 package get
 
 import (
-	"fmt"
 	"os"
+	"path"
 	"path/filepath"
+	"strings"
 
 	"github.com/GoogleContainerTools/kpt/internal/util/fetch"
-	"github.com/GoogleContainerTools/kpt/internal/util/git"
 	"github.com/GoogleContainerTools/kpt/internal/util/pkgutil"
 	"github.com/GoogleContainerTools/kpt/internal/util/stack"
 	kptfilev1alpha2 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1alpha2"
 	"github.com/GoogleContainerTools/kpt/pkg/kptfile/kptfileutil"
 	"sigs.k8s.io/kustomize/kyaml/errors"
-	"sigs.k8s.io/kustomize/kyaml/pathutil"
 )
 
 // Command fetches a package from a git repository, copies it to a local
 // directory, and expands any remote subpackages.
 type Command struct {
 	// Git contains information about the git repo to fetch
-	kptfilev1alpha2.GitLock
+	Git *kptfilev1alpha2.Git
 
 	// Destination is the output directory to clone the package to.  Defaults to the name of the package --
 	// either the base repo name, or the base subdirectory name.
 	Destination string
 
 	// Name is the name to give the package.  Defaults to the destination.
 	Name string
-
-	// Remove directory before copying to it.
-	Clean bool
 }
 
 // Run runs the Command.
 func (c Command) Run() error {
-	revertFunc, err := c.updateParentKptfile()
+	if err := (&c).DefaultValues(); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(c.Destination); !os.IsNotExist(err) {
+		return errors.Errorf("destination directory %s already exists", c.Destination)
+	}
+
+	err := os.MkdirAll(c.Destination, 0700)
+	if err != nil {
+		return err
+	}
+
+	// normalize path to a filepath
+	repoDir := c.Git.Directory
+	if !strings.HasSuffix(repoDir, "file://") {
+		repoDir = filepath.Join(path.Split(repoDir))
+	}
+	c.Git.Directory = repoDir
+
+	kf := kptfileutil.DefaultKptfile(c.Name)
+	kf.Upstream = &kptfilev1alpha2.Upstream{
+		Type:           kptfilev1alpha2.GitOrigin,
+		Git:            c.Git,
+		UpdateStrategy: "resource-merge",
+	}
+
+	err = kptfileutil.WriteFile(c.Destination, kf)
```

Comment:
Interesting... so we write a default Kptfile with upstream info, and then fetch does the rest and merged the kptfile.
This is more of a note for myself.

Reply:
Yeah, the idea is to do this in "two phases". First update the desired state and then call the functionality for "reconciling" the actual state (in this case the package on disk) with the desired state. This supports use-cases where the desired state is updated by other tools or by users manually, or by kpt porcelain cli commands.